flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegate.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class NettyMessageClientDecoderDelegate extends ChannelInboundHandlerAdapter {
+
+	/** The decoder for BufferResponse. */
+    private final NettyMessageDecoder bufferResponseDecoder;
+
+    /** The decoder for messages other than BufferResponse. */
+	private final NettyMessageDecoder nonBufferResponseDecoder;
+
+	/** The accumulation buffer for the frame header. */
+	private ByteBuf frameHeaderBuffer;
+
+	/** The decoder for the current message. It is null if we are decoding the frame header. */
+	private NettyMessageDecoder currentDecoder;
+
+    NettyMessageClientDecoderDelegate(NetworkClientHandler networkClientHandler) {
+		this.bufferResponseDecoder = new BufferResponseDecoder(
+			new NetworkBufferAllocator(
+				checkNotNull(networkClientHandler)));
+        this.nonBufferResponseDecoder = new NonBufferResponseDecoder();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        bufferResponseDecoder.onChannelActive(ctx);
+        nonBufferResponseDecoder.onChannelActive(ctx);
+
+		frameHeaderBuffer = ctx.alloc().directBuffer(FRAME_HEADER_LENGTH);
+
+		super.channelActive(ctx);
+    }
+
+	/**
+	 * Releases resources when the channel is closed. When exceptions are thrown during processing received netty
+	 * buffers, {@link CreditBasedPartitionRequestClientHandler} is expected to catch the exception and close the
+	 * channel and trigger this notification.
+	 *
+	 * @param ctx The context of the channel close notification.
+	 */
+	@Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+		bufferResponseDecoder.close();
+		nonBufferResponseDecoder.close();
+
+		frameHeaderBuffer.release();
+
+		super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+    	if (!(msg instanceof ByteBuf)) {
+			ctx.fireChannelRead(msg);
+			return;
+		}
+
+        ByteBuf data = (ByteBuf) msg;
+        try {
+            while (data.isReadable()) {
+            	if (currentDecoder != null) {
+					NettyMessageDecoder.DecodingResult result = currentDecoder.onChannelRead(data);
+					if (!result.finished) {
+						break;
+					}
+					ctx.fireChannelRead(result.message);
+
+					currentDecoder = null;
+					frameHeaderBuffer.clear();
+				}
+
+				decodeFrameHeader(data);
+            }
+            checkState(!data.isReadable(), "Not all data of the received buffer consumed.");
+        } finally {
+            data.release();
+        }
+    }
+
+    private void decodeFrameHeader(ByteBuf data) {
+		ByteBuf toDecode = ByteBufUtils.accumulate(
+			frameHeaderBuffer,
+			data,
+			FRAME_HEADER_LENGTH,
+			frameHeaderBuffer.readableBytes());
+		if (toDecode != null) {
```

Comment:
nit: insert an empty line before `if`, because too many lines close together.

Reply:
Fixed the format.